test/physics/em/Rayleigh.test.cc

```diff
@@ -33,46 +46,151 @@ class RayleighInteractorTest : public celeritas_test::InteractorHostTestBase
+        };
+        this->set_material_params(inp);
+        this->set_material("PbWO");
+
+        // Construct RayleighModel and set the host data group
+        model_ = std::make_shared<RayleighModel>(
+            ModelId{0}, particles, *this->material_params());
+        group_ = model_->host_group();
     }
 
     void sanity_check(const Interaction& interaction) const
     {
         ASSERT_TRUE(interaction);
 
-        // Check change to parent track
-        EXPECT_GT(this->particle_track().energy().value(),
+        // Check change to parent track - coherent scattering
+        EXPECT_EQ(this->particle_track().energy().value(),
                   interaction.energy.value());
-        EXPECT_LT(0, interaction.energy.value());
-        EXPECT_SOFT_EQ(1.0, celeritas::norm(interaction.direction));
         EXPECT_EQ(celeritas::Action::scattered, interaction.action);
-
-        // XXX Check secondaries
-
-        // Check conservation between primary and secondaries
-        this->check_conservation(interaction);
     }
 
   protected:
-    celeritas::detail::RayleighInteractorPointers pointers_;
+    std::shared_ptr<RayleighModel>       model_;
+    celeritas::detail::RayleighNativeRef group_;
 };
 
 //---------------------------------------------------------------------------//
 // TESTS
 //---------------------------------------------------------------------------//
 
-TEST_F(RayleighInteractorTest, basic) {}
+TEST_F(RayleighInteractorTest, basic)
+{
+    // Sample an element (TODO: add ElementSelector)
+    ElementId el_id{0};
+
+    std::vector<double>            angle;
+    std::vector<unsigned long int> rng_counts;
+
+    // Sample scattering angle and count rng used for each incident energy
+    for (double inc_e : {0.05, 0.1, 1.0, 10.0, 100.0})
+    {
+        // Set the incident particle energy
+        this->set_inc_particle(pdg::gamma(), MevEnergy{inc_e});
+
+        // Create the interactor
+        RayleighInteractor interact(this->model_->host_group(),
+                                    this->particle_track(),
+                                    this->direction(),
+                                    el_id);
+
+        RandomEngine& rng_engine = this->rng();
+
+        // Produce a sample from the original/incident photon
+        celeritas::Interaction result = interact(rng_engine);
+        SCOPED_TRACE(result);
+        this->sanity_check(result);
+        rng_engine.count();
+
+        angle.push_back(dot_product(result.direction, this->direction()));
+        rng_counts.push_back(rng_engine.count());
+    }
+
+    const double expected_angle[] = {-0.726848858395344,
+                                     -0.95887836792401,
+                                     -0.910276761243175,
+                                     -0.223090692143936,
+                                     -0.388188923742863};
+
+    const unsigned long int expected_rng_counts[]
+        = {17774ul, 19342ul, 19416ul, 19424ul, 19438ul};
```

Comment:
Holy hell, a single interaction take 17K random numbers (at the low end of the energy range)?  It looks like at higher energies it's more reasonable, but could this indicate a bug in the rejection sampling? Also since the model is valid down to 1e-5 MeV (our test geant data has cross sections down to 1e-4) it might be worth testing an even lower energy to see if this explodes.

Reply:
After putting reset_count, now average rng counts are
<pre>
{17774ul, 1568ul, 74ul, 8ul, 14ul};
{49710ul, 11914ul, 137ul, 12ul, 10ul};
</pre>
Agrees that there is something wrong in the low energy reason, which I was planning to investigate today and validate results with respect to Geant4 today.  As you noticed, I did not try at low energy as the test was hanging very long time, apparently stuck in the rng generation from debugging (but could be false information).  Anyway, I will spend sometime for validation.

Reply:
Hah I believe that following the trend you'd see an apparent hang at lower energies :) OK, let me know if you want to keep debugging on this branch, or merge it in and do validation in a separate branch. If the former I can take a closer look at the sampling implementation and compare against the reference documentation and the geant implementation.

Reply:
@sethrj I will validate results today with respect to Geant4 and may understand the low sampling efficiency.  Since it smells like a bug (or involve a wrong unit conversion), please hold the merge.  Of course, you are welcome to take a closer look at the geant4 code, but busy for other stuffs.  If I need a help, I will let you know.